pkgs/build-support/testers/test/default.nix

```
@@ -99,117 +99,116 @@ lib.recurseIntoAttrs {
-        ) || {
-          echo "Test failed: could not find pattern in build log $log"
-          exit 1
-        }
-        echo 'All good.'
-        touch $out
+      actual = runCommand "actual" { } ''
+        echo test >"$out"
+        chmod a+x -- "$out"
       '';
+    };
 
-    fileMissing =
-      runCommand "testEqualContents-fileMissing" {
-        log = testers.testBuildFailure (testers.testEqualContents {
-          assertion = "The same directory contents at different paths are recognized as equal";
-          expected = runCommand "expected" {} ''
-            mkdir -p $out/c
-            echo a >$out/a
-            echo b >$out/b
-            echo d >$out/c/d
-          '';
-          actual = runCommand "actual" {} ''
-            mkdir -p $out/c
-            echo a >$out/a
-            echo d >$out/c/d
-          '';
-        });
-      } ''
+    unequalExe = testers.testBuildFailure (
+      testers.testEqualContents {
+        assertion = "Different file mod bits are not recognized as equal";
+        expected = runCommand "expected" { } ''
+          touch -- "$out"
+          chmod a+x -- "$out"
+        '';
+        actual = runCommand "actual" { } ''
+          touch -- "$out"
+        '';
+      }
+    );
+
+    unequalExeInDir = testers.testBuildFailure (
+      testers.testEqualContents {
+        assertion = "Different file mod bits are not recognized as equal in directory";
+        expected = runCommand "expected" { } ''
+          mkdir -p -- "$out/a"
+          echo b >"$out/b"
+          chmod a+x -- "$out/b"
+        '';
+        actual = runCommand "actual" { } ''
+          mkdir -p -- "$out/a"
+          echo b >"$out/b"
+        '';
+      }
+    );
+
+    fileDiff =
+      let
+        log = testers.testBuildFailure (
+          testers.testEqualContents {
+            assertion = "Different files in are not recognized in equal in directories";
+            expected = runCommand "expected" { } ''
+              mkdir -p -- "$out/b"
+              echo a >"$out/a"
+              echo EXPECTED >"$out/b/c"
+            '';
+            actual = runCommand "actual" { } ''
+              mkdir -p "$out/b"
+              echo a >"$out/a"
+              echo ACTUAL >"$out/b/c"
+            '';
+          }
+        );
+      in
+      runCommand "testEqualContents-fileDiff" { inherit log; } ''
+        cat "$log/testBuildFailure.log"
         (
           set -x
-          grep -F -- "Contents must be equal but were not" $log/testBuildFailure.log
-          grep -E -- 'Only in .*-expected: b' $log/testBuildFailure.log
+          grep -F -- "Contents must be equal, but were not!" "$log/testBuildFailure.log" &&
+          grep -E -- '\+\+\+ .*-expected/b/c' "$log/testBuildFailure.log" &&
+          grep -E -- '--- .*-actual/b/c' "$log/testBuildFailure.log" &&
+          grep -F -- "-ACTUAL" "$log/testBuildFailure.log" &&
+          grep -F -- "+EXPECTED" "$log/testBuildFailure.log"
```

```suggestion
          grep -F -- "Contents must be equal, but were not!" "$log/testBuildFailure.log"
          grep -E -- '\+\+\+ .*-expected/b/c' "$log/testBuildFailure.log"
          grep -E -- '--- .*-actual/b/c' "$log/testBuildFailure.log"
          grep -F -- "-ACTUAL" "$log/testBuildFailure.log"
          grep -F -- "+EXPECTED" "$log/testBuildFailure.log"
```

Aren't we in `set -e` mode, `errexit`?

---

>Aren't we in `set -e` mode, `errexit`?

And even `inherit_errexit` from stdenv. Bash is being grumpy. This is either a bug or a known quirk, I’m not really sure—`dash` has the same behavior.

- Case 1, works as expected.
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 1.
  (
  	set -x

  	false

  	# OK, this is not reachable.
  	echo unreachable
  )
  ```
  ```
  + false
  ```

- Case 2 (old test). Never trust errexit when pipes or subshells are involved.
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 2.
  (
  	set -x
  	# optionally, explicitly set errexit
  	# (e.g. dash does not support setting inherit_errexit via shopt)
  	set -e

  	false

  	echo FAIL
  ) || {
  	echo oh no
  }
  ```
  ```
  + set -e
  + false
  + echo FAIL
  FAIL
  ```

- Case 3 (updated test).
  ```bash
  set -e -u
  shopt -s inherit_errexit

  # Case 3.
  (
  	set -x
  	false && echo OK
  ) || {
  	echo fails as expected
  }
  ```
  ```
  + false
  fails as expected
  ```

---

Added a comment to explain that.

https://github.com/NixOS/nixpkgs/blob/b40d043d5a2a1fb473f1c6c853dac444ceadab62/pkgs/build-support/testers/test/default.nix#L216-L219